File: Different_Ways_to_Add_Parentheses/Different_Ways_to_Add_Parentheses.go
package main

import (
	"fmt"
	"strconv"
)

func main() {
	expression := "2-1-1"
	fmt.Println(diffWaysToCompute(expression))
}

func diffWaysToCompute(expression string) []int {
	output := make([]int, 0)
	for i := 0; i < len(expression); i++ {
		if expression[i] == '-' || expression[i] == '+' || expression[i] == '*' {
			left := diffWaysToCompute(string(expression[:i]))
			right := diffWaysToCompute(string(expression[i+1:]))
			for _, x := range left {
				for _, y := range right {
					if expression[i] == '+' {
						output = append(output, x+y)
					} else if expression[i] == '-' {
						output = append(output, x-y)
					} else {
						output = append(output, x*y)
					}
				}
			}

		}
	}

	if len(output) == 0 {
		x, _ := strconv.Atoi(expression)
		output = append(output, x)
	}
	return output
}

File: Different_Ways_to_Add_Parentheses/Different_Ways_to_Add_Parentheses.rs
fn diffWaysToCompute(expression:String) -> Vec<i32>{
    let mut v1:Vec<i32> = Vec::new();
    
    for (i,s) in expression.chars().enumerate(){
        if s == '+' || s == '-' || s == '*'{
            let left = diffWaysToCompute((&expression[0..i]).to_string());
            let right = diffWaysToCompute((&expression[i+1..]).to_string());
            for x in left.iter(){
                for y in right.iter(){
                    match s{
                        '+' => v1.push(x+y),
                        '-' => v1.push(x-y),
                        '*' => v1.push(x*y),
                        _ => {}
                    }
                }
            }
        }
    }
    if v1.is_empty(){
        v1.push(expression.parse::<i32>().unwrap())
    }
    return v1
}

fn main(){
    let expression = "2-1-1".to_string();
    println!("{:?} ",diffWaysToCompute(expression));
}